ironfist/src/cpp/scripting/scripting.cpp

```diff
@@ -772,41 +772,156 @@ char *GetScriptContents() {
+}
+
+void ErrorSavingMapVariable(std::string& mapVariableId, const std::string& addErrorMessage) {
+	const std::string s2("' could not be saved properly.");
+	ErrorMapVariable(mapVariableId, s2, addErrorMessage);
+}
+
+void ErrorLoadingMapVariable(std::string& mapVariableId, const std::string& addErrorMessage) {
+	const std::string s2("' could not be loaded properly.");
+	ErrorMapVariable(mapVariableId, s2, addErrorMessage);
+}
+
+bool isTable(MapVarType type) {
+	return (type == MAPVAR_TYPE_TABLE);
+}
+
+bool isStringNumBool(MapVarType type) {
+	return (type == MAPVAR_TYPE_STRING || type == MAPVAR_TYPE_NUMBER || type == MAPVAR_TYPE_BOOLEAN);
+}
+
+MapVarType StringToMapVarType(std::string stringType) {
+	if (stringType == "string") {
+		return MAPVAR_TYPE_STRING;
+	} else if (stringType == "number") {
+		return MAPVAR_TYPE_NUMBER;
+	} else if (stringType == "boolean") {
+		return MAPVAR_TYPE_BOOLEAN;
+	} else if (stringType == "table") {
+		return MAPVAR_TYPE_TABLE;
+	} else {
+		return MAPVAR_TYPE_ERROR;
+	}
+}
+
+std::string getMVKeyFromLUA() {
+	MapVarType type = StringToMapVarType(lua_typename(map_lua, lua_type(map_lua, -2)));
+	if (type == MAPVAR_TYPE_STRING) {
+		std::string stringValue(lua_tostring(map_lua, -2));
+		return stringValue;
+	} else if (type == MAPVAR_TYPE_NUMBER) {
+		return std::to_string(lua_tonumber(map_lua, -1));
+	}
+}
+
+std::string GetMVValueFromLUA(MapVarType &type) {
+	if (type == MAPVAR_TYPE_STRING) {
+		std::string stringValue(lua_tostring(map_lua, -1));
+		return stringValue;
+	} else if (type == MAPVAR_TYPE_NUMBER) {
+		return std::to_string(lua_tonumber(map_lua, -1));
+	} else if (type == MAPVAR_TYPE_BOOLEAN) {
+		return std::to_string(lua_toboolean(map_lua, -1));
+	}
+}
+
+void GetMVTablesFromLUA(luaTable &lt, std::string &mapVariableId) {
+	luaTableElements lte;
+	luaSubTables lst;
+	lua_pushnil(map_lua);
+	while (lua_next(map_lua, -2) != 0) {
+		MapVarType valueType = StringToMapVarType(lua_typename(map_lua, lua_type(map_lua, -1)));
+		std::string key = getMVKeyFromLUA();
+		if (isStringNumBool(valueType)) {
+			lte[key] = std::pair<MapVarType, std::string>(valueType, GetMVValueFromLUA(valueType));
+			lua_pop(map_lua, 1);
+		} else if (isTable(valueType)) {
+			lst.push_back(key);
+			GetMVTablesFromLUA(lt, key);
+		} else {
+			ErrorSavingMapVariable(mapVariableId, " Wrong type in the table.");
+		}
+	}
+	lt[mapVariableId] = std::pair<luaTableElements, luaSubTables>(lte, lst);
+	lua_pop(map_lua, 1);
+}
+
+std::map<std::string, mapVariable> LoadMapVariablesFromLUA() {
+
+	std::map<std::string, mapVariable> mapVariables;
+	lua_getglobal(map_lua, "mapVariables");
+
+	if (lua_isnil(map_lua, -1)) return mapVariables;
+
+	lua_pushnil(map_lua);
+	while (lua_next(map_lua, -2) != 0) {
+		std::string mapVariableId(lua_tostring(map_lua, -1));
+		lua_getglobal(map_lua, mapVariableId.c_str());
+		MapVarType mapVariableType = StringToMapVarType(lua_typename(map_lua, lua_type(map_lua, -1)));
+		mapVariable mapVar;

```

Comment:
Note that this is a local variable within the while loop. That means that each version of mapVar for each iteration of the loop will be the same. If you evaluated &mapVar somewhere, that value would be the same for each iteration of the loop.

All this is fine because the mapVariables map stores a mapVariable, not a mapVariable* . That means that, when you write into mapVariables on line 881, it will actually be copied into the map.